go-controller/pkg/node/gateway_shared_intf.go

```diff
@@ -245,16 +264,73 @@ func newSharedGatewayOpenFlowManager(patchPort, macAddress, gwBridge, gwIntf str
 		dftFlows = append(dftFlows,
 			fmt.Sprintf("cookie=%s, priority=50, in_port=%s, ipv6, "+
 				"actions=ct(zone=%d, table=1)", defaultOpenFlowCookie, ofportPhys, config.Default.ConntrackZone))
+
+		physicalIP, err := util.MatchIPNetFamily(true, ips)
+		if err != nil {
+			return nil, fmt.Errorf("unable to determine IPv6 physical IP of host: %v", err)
+		}
+		// table 0, SVC Hairpin from OVN destined to local host, DNAT to host, send to table 4
+		dftFlows = append(dftFlows,
+			fmt.Sprintf("cookie=%s, priority=500, in_port=%s, ipv6, ipv6_dst=%s, ipv6_src=%s,"+
+				"actions=ct(commit,zone=%d,nat(dst=%s),table=3)",
+				defaultOpenFlowCookie, ofportPatch, types.V6HostMasqueradeIP, physicalIP.IP,
+				HostMasqCTZone, physicalIP.IP))
+
+		// table 0, Reply SVC traffic from Host -> OVN, unSNAT and goto table 5
+		dftFlows = append(dftFlows,
+			fmt.Sprintf("cookie=%s, priority=500, in_port=LOCAL, ipv6, ipv6_dst=%s,"+
+				"actions=ct(zone=%d,nat,table=5)",
+				defaultOpenFlowCookie, types.V6OVNMasqueradeIP, OVNMasqCTZone))
 	}
 
-	// table 1, established and related connections go to pod
-	dftFlows = append(dftFlows,
-		fmt.Sprintf("cookie=%s, priority=100, table=1, ct_state=+trk+est, "+
-			"actions=output:%s", defaultOpenFlowCookie, ofportPatch))
+	var protoPrefix string
+	var masqIP string
 
-	dftFlows = append(dftFlows,
-		fmt.Sprintf("cookie=%s, priority=100, table=1, ct_state=+trk+rel, "+
-			"actions=output:%s", defaultOpenFlowCookie, ofportPatch))
+	// table 0, packets coming from Host -> Service
```

Comment:
Question: couldn't we apply this to node port services as well and get rid of the shared gateway iptables rules, which only exist for the purpose of DNAT-ing Host -> node port service -> cluster IP?

Reply:
nice observation. Yes we could do this. I guess it wouldn't be much of a perf gain, but it would consolidate things more to OpenFlow. Let's keep this comment and we can make a github issue and follow up on it. Sound good?

Reply:
Actually, I just realized we need iptables for nodeport to the same node, otherwise kernel will try to deliver it locally on the host, and we want it to go to OVN.

Reply:
Ah, yes we only need iptables for node port services being targeted by something on the host itself. No node port related traffic crosses nodes (though it might when https://github.com/ovn-org/ovn-kubernetes/issues/1951 is implemented). But OK for what concerns my original comment.